### ISSUE TITLE
Request: ability to run post-snapshot hook regardless of result of pre-snapshot hook

### ISSUE THREAD

**Issue by helamonster** [zrepl/zrepl#392]
The docs say:

> Jobs with periodic snapshots can run hooks before and/or after taking the snapshot specified in snapshotting.hooks: Hooks are called per filesystem before and after the snapshot is taken (pre- and post-edge). Pre-edge invocations are in configuration order, post-edge invocations in reverse order, i.e. like a stack. If a pre-snapshot invocation fails, err_is_fatal=true cuts off subsequent hooks, does not take a snapshot, and only invokes post-edges corresponding to previous successful pre-edges. err_is_fatal=false logs the failed pre-edge invocation but does not affect subsequent hooks nor snapshotting itself. Post-edges are only invoked for hooks whose pre-edges ran without error. Note that hook failures for one filesystem never affect other filesystems.

So it sounds like if a pre-edge hook fails for **any** reason, the post-edge hook will **never** run. 
There should be a way to still run the post-edge hook even if the pre-edge hook fails (maybe with the error code / message from the failure).

My use case:

pre-snapshot hook executes:
`xfs_freeze -f /filesystem`

post-snapshot hook executes:
`xfs_freeze -u /filesystem`

The pre hook is sometimes timing out but still freezing the filesystem. I need to **ALWAYS** execute the post hook to ensure the filesystem is never left frozen. Is there a way I can accomplish this now, or would this require code changes?

**Comment by helamonster** [comment]
See also: https://lwn.net/Articles/287435/

The xfs_freeze CLI program doesn't appear to support the timeout option, which ideally would be used in the ioctl freeze call:
```
int timeoutSec = 30;
ioctl(fd, FIFREEZE, &timeoutSec);
```

Even with a later call to:
```
ioctl(fd, FITHAW, 0);
```
There's still a chance that the thaw never gets executed, leaving the filesystem frozen.

Perhaps a new hook type could be implemented that handles all of this logic rather than shelling out to a script that executes xfs_freeze or fsfreeze?

**Comment by problame** [comment]
> So it sounds like if a pre-edge hook fails for any reason, the post-edge hook will never run.

Yes, but nothing stops you from just doing something like the following

```bash
# /etc/zrepl/zfs_freeze.zrepl.hook.bash

set -euo pipefail

case "$ZREPL_HOOKTYPE" in
   pre_snapshot)
     xfs_freeze -f /filesystem || true
    ;;
  post_snapshot)
     xfs_freeze -u /filesystem
    ;;
esac
```
(And probably just increase the hook timeout?)

> The pre hook is sometimes timing out but still freezing the filesystem

That's ... weird behavior from `xfs_freeze` I suppose, it seems like XFS freezing is not atomic? 

Btw: why are you xfs_freezing anything in a ZFS snapshot hook?

**Comment by helamonster** [comment]
Yeah, the freeze does have to sync data to disk and due to heavy I/O in my case it was taking minutes in some cases.
I had pretty much the same thing going on in your example hook script with the timeout set to 30 seconds. But it wasn't stopping after that timeout.. I'm not sure how zrepl handles the timeout in regards to killing the hook script but my guess is that once the xfs_freeze starts it has to complete (can't be simply aborted by killing the process). So even if the process is killed, the snapshot taken, then the post snapshot xfs_freeze -u is executed, it too may have to first wait for the freeze to finish.

I'll try just increasing the timeout and see how that goes but I would still like to be able to know if the pre snapshot hook times out.
I suppose I could use the "timeout" command with a value lower than the zrepl snapshot hook timeout.

**Comment by helamonster** [comment]
As for why I am freezing XFS in a ZFS snapshot hook... Unfortunately I have an application that requires XFS (for the reflink support). So I'm using a ZVOL for the underlying block device of the XFS filesystem.

It looks like there's some work being done by Pawel Dawidek with Fudo Security to implement reflink support for ZFS <https://openzfs.org/wiki/OpenZFS_Developer_Summit_2020_talks#File_Cloning_with_Block_Reference_Table_.28Pawel_Dawidek.29>, but until then, XFS is the only option. Of course, it doesn't have to be on top of ZFS, but I want those snapshots ;-)

**Comment by problame** [comment]
> I suppose I could use the "timeout" command with a value lower than the zrepl snapshot hook timeout.

Yeah, you could use that command to make the hook exit with 0 status code.

However, you should ~~think about~~ investigate what it means to make a zfs snapshot of a zvol where the XFS volume on top of it is still _in the process of freezing_. Because that's what's happening right now. I'd assume the snapshots don't actually capture a fully frozen XFS filesystem but some intermediary state. 

> 
> As for why I am freezing XFS in a ZFS snapshot hook... Unfortunately I have an application that requires XFS (for the reflink support). So I'm using a ZVOL for the underlying block device of the XFS filesystem.
> 
> It looks like there's some work being done by Pawel Dawidek with Fudo Security to implement reflink support for ZFS https://openzfs.org/wiki/OpenZFS_Developer_Summit_2020_talks#File_Cloning_with_Block_Reference_Table_.28Pawel_Dawidek.29, but until then, XFS is the only option. Of course, it doesn't have to be on top of ZFS, but I want those snapshots ;-)

Interesting, I'm sure PJD (and myself as well :D) would be interested in what application that is ;)

**Comment by helamonster** [comment]
> However, you should think about investigate what it means to make a zfs snapshot of a zvol where the XFS volume on top of it is still in the process of freezing. Because that's what's happening right now. I'd assume the snapshots don't actually capture a fully frozen XFS filesystem but some intermediary state.

If that's what's happening, then I certainly wouldn't trust the consistency of the XFS filesystem at the time of the ZFS snapshot.

At least when the `xfs_freeze -f` command is not timing out it does seem to result in a consistent XFS filesystem snapshot as I would expect. I've been testing by cloning the snapshot of the ZVOL and using `xfs_repair` to check for issues. When the freeze was timing out, I wasn't sure exactly what was happening behind the scenes as far as the timing of the zfs snapshot (but it didn't seem like zrepl was actually killing the hook script after the timeout). As you can imagine, I don't want that to ever happen.

> Interesting, I'm sure PJD (and myself as well :D) would be interested in what application that is ;)

The application is Veeam. I have a Linux backup repository serving as the storage backend. It only supports reflinks on XFS at the moment. I would looove to be able to use that natively on ZFS and not have to mess with XFS or ZVOLs.

@pjd I'd love to be in the loop on the progress. And if you need any testers, let me know :-)
